src/core/alloc.h

```diff
@@ -68,12 +68,51 @@ MVM_STATIC_INLINE void MVM_free(void *p) {
     (addr) = NULL; \
 } while (0)
 
-MVM_STATIC_INLINE void MVM_free_at_safepoint(MVMThreadContext *tc, void *ptr) {
-    MVM_VECTOR_PUSH(tc->instance->free_at_safepoint, ptr);
+/* Entry in the "free at safe point" linked list. */
+struct MVMAllocSafepointFreeListEntry {
+    void                           *to_free;
+    MVMAllocSafepointFreeListEntry *next;
+};
+
+/* Race to add a piece of memory to be freed at the next global safe point.
+ * A global safe point is defined as one in which all threads, since we
+ * requested the freeing of the memory, have reached a safe point. */
+MVM_STATIC_INLINE void MVM_free_at_safepoint(MVMThreadContext *tc, void *to_free) {
+    MVMAllocSafepointFreeListEntry *orig;
+    MVMAllocSafepointFreeListEntry *to_add = MVM_malloc(sizeof(MVMAllocSafepointFreeListEntry));
+    to_add->to_free = to_free;
+    do {
+        orig = tc->instance->free_at_safepoint;
+        to_add->next = orig;
+    } while (!MVM_trycas(&(tc->instance->free_at_safepoint), orig, to_add));
 }
 
+MVM_STATIC_INLINE void * MVM_realloc_at_safepoint(MVMThreadContext *tc, void *p, size_t old_bytes, size_t new_bytes) {
+    void *allocd;
+#ifdef MVM_USE_MIMALLOC
+    allocd = mi_expand(p, new_bytes);
```

Comment:
A comment pointing out that mi_expand tries to expand in-place and therefore is safe to do right away (and not at a safepoint) would be nice.

Reply:
Done, rebased (plus some typos fixed in the commit message), and force pushed.